### ISSUE TITLE
Parsing state doesnt propagate `serde_json::Error`

### ISSUE THREAD

**Issue by Keksoj** [sozu-proxy/sozu#744]
We kept getting the same log  

```
error consuming load state message
```

without knowing why. A heroic debug session by @Keruspe showed that the bug lies in the `parse()` function, that relies on [nom](https://github.com/Geal/nom), and uses [serde_json](https://docs.serde.rs/serde_json/index.html) to parse the state file into `sozu_command_lib::CommandRequest` :

```rust
use nom::{
    bytes::streaming::is_not,
    character::streaming::char,
    combinator::{complete, map_res},
    multi::many0,
    sequence::terminated,
};

pub fn parse(input: &[u8]) -> IResult<&[u8], Vec<CommandRequest>> {
    use serde_json::from_slice;
    many0(complete(terminated(
        map_res(is_not("\0"), from_slice),
        char('\0'),
    )))(input)
}
```

A trivial [serde_json::Error](https://docs.serde.rs/serde_json/struct.Error.html) such as this  :

```
Serde error: Error("missing field `address`", line: 1, column: 7688)
```

is NOT propagated. This explains:

- #731 

 We HAVE to propagate it, to log it on the server and return to the CLI.

**Comment by Keksoj** [comment]
Thoughts @geal ?

**Comment by Geal** [comment]
To propagate the error, you can use a custom error type that implements FromExternalError with serde_json errors: https://docs.rs/nom/7.1.0/nom/combinator/fn.map_res.html
The map_res combinator is designed for that.
You will also need the [cut](https://docs.rs/nom/7.1.0/nom/combinator/fn.cut.html) combinator to transform the Error into a Failure, then the many0 combinator will return an error instead of stopping silently

**Comment by Geal** [comment]
How did it happen though? Were you using different versions of the command lib on the client and sozu sides?

**Comment by Keksoj** [comment]
Same version, the `0.14` dev branch. `PROTOCOL_VERSION` is the same, 0. But the `state.json` we wanted to load came from a `0.13` version on Sōzu I believe.

Thanks for your pinpointing into the relevant functions to solve the issue :+1:

**Comment by Keksoj** [comment]
Hello @Geal , I did my best to reproduce the problem here : https://github.com/Keksoj/propagate_error_on_nom

But I didn't manage:

- to create a Custom error to wrap serde_json error,
- to implement `FromExternalError`
- to use `cut`

My feeling is that I lack examples of custom errors implementing `FromExternalError`, and that I still have a limited understanding of how `map_res` deals with `cut`.
Would you mind having a look at my sandbox repo and commenting with hints somewhere?

**Comment by Geal** [comment]
`FromExternalError` can do whatever you want, its job is to take the error returned by the function in map_res to your error type. If the error type is `(Input, ErrorKind)` it will just drop it.
`cut` does not interact with `map_res`. The combinator returns an `Err::Error(_)`, `cut` transforms an `Err::Error(_)` to `Err::Failure(_)`( and lets anything else pass through).
`many0` applies a parser repeatedly until it returns an error. If it was an `Err::Error(_)` it will return a `Vec<_>` of accumulated errors, while for `Err::Failure(_)` it will return the failure.

That's why cut is used here: to differentiate between an error that indicates the end (with complete and terminated) from an error in json deserialization (with map_res)

**Comment by Keksoj** [comment]
Closed by #747
